### PR TITLE
Gallery example showing LOS intersections with simulation box

### PR DESCRIPTION
**Pull request by wtbarnes**
This PR adds a gallery example showing how to compute intersections between observer LOS and a simulation domain defined by a cartesian box.

### CODE REVIEW

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+"""
+============================================================================================
+Understanding Observer Orientation Relative to a Simulated Volume with `astropy.coordinates`
+============================================================================================
+
+This notebook demonstrates how to use the `astropy.coordinates` framework, combined with the solar coordinate frameworks provided by SunPy, to find the intersection of a set of lines of sight defined by an observer position with a simulated volume represented by a Cartesian box.
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:6]:
```suggestion
This example demonstrates how to use the `astropy.coordinates` framework, combined with the solar
coordinate frameworks provided by ``sunpy``, to find the intersection of a set of lines of sight
defined by an observer position with a simulated volume represented by a Cartesian box.
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import sunpy.map
+import sunpy.sun.constants
+from sunpy.coordinates import Heliocentric, Helioprojective
+
+#############################################################################################
+# First, we need to define the orientation of our box and the corresponding coordinate frame.
+# We do this by specifying a coordinate in the Heliographic Stonyhurst frame (HGS) and then
+# using that coordinate to define the coordinate frame of the simulation box in a
+# Heliocentric Cartesian (HCC) frame. For more information on all of these frames, see
+# `Thompson (2006) <https://ui.adsabs.harvard.edu/abs/2006A%26A...449..791T/abstract>`_.
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:28]:
```suggestion
# `Thompson (2006) <https://ui.adsabs.harvard.edu/abs/2006A%26A...449..791T/abstract>`__.
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+# simulation is centered on this point.
+
+def make_fake_map(ref_coord):
+    instrument_data = np.nan * np.ones((25, 25))
+    instrument_header = sunpy.map.make_fitswcs_header(instrument_data,
+                                                      ref_coord,
+                                                      scale=u.Quantity([20, 20])*u.arcsec/u.pix)
+    return sunpy.map.Map(instrument_data, instrument_header)
+instrument_map = make_fake_map(hcc_orientation.transform_to(frame_hpc))
+
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:59]:
```suggestion
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+#############################################################################################
+# Now that we have our map representing our fake observation, we can get the coordinates, in
+# pixel space, of the center of each pixel. These will be our lines of sight.
+
+map_indices = sunpy.map.all_pixel_indices_from_map(instrument_map).value.astype(int)
+map_indices = map_indices.reshape((2, map_indices.shape[1]*map_indices.shape[2]))
+
+#############################################################################################
+# We can then use the WCS of the map to find the associate world coordinate for each pixel coordinate. Note that by default, the "z" or *distance* coordinate in the HPC frame is assumed to lie on the solar surface. As such, for each LOS, we will add a z coordinate that spans from 99% to 101% of the observer radius. This gives us a reasonable range of distances that will intersect our simulation box.
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:69]:
```suggestion
# We can then use the WCS of the map to find the associate world coordinate for each pixel coordinate.
# Note that by default, the "z" or *distance* coordinate in the HPC frame is assumed to lie on the solar surface.
# As such, for each LOS, we will add a z coordinate that spans from 99% to 101% of the observer radius.
# This gives us a reasonable range of distances that will intersect our simulation box.
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+#############################################################################################
+# We can do a simple visualization of all of our LOS on top of our synthetic map
+
+fig = plt.figure()
+ax = fig.add_subplot(projection=instrument_map)
+instrument_map.plot(axes=ax, title=False)
+instrument_map.draw_grid(axes=ax, color='k')
+for los in lines_of_sight:
+    ax.plot_coord(los[0], color='C0', marker='.', ls='', markersize=1,)
+plt.show()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:86]:
```suggestion
```

One show at the end if i recall.

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+box_origin = hcc_orientation.transform_to(frame_hcc)
+box_origin = SkyCoord(x=box_origin.x,
+                      y=box_origin.y,
+                      z=box_origin.z+box_dimensions[2]/2,
+                      frame=box_origin.frame)
+
+#############################################################################################
+# Using that origin, we can compute the coordinates of all edges.
+edge_coords = []
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:120]:
```suggestion

edge_coords = []
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+for edge in edges:
+    edge_coords.append(SkyCoord(x=box_origin.x+u.Quantity([edge[0][0],edge[1][0]]),
+                                y=box_origin.y+u.Quantity([edge[0][1],edge[1][1]]),
+                                z=box_origin.z+u.Quantity([edge[0][2],edge[1][2]]),
+                                frame=box_origin.frame))
+
+#############################################################################################
+# Let's overlay the simulation box on top of our fake map we created earlier to see if things
+# look right.
+fig = plt.figure()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:130]:
```suggestion

fig = plt.figure()
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+ax = fig.add_subplot(projection=instrument_map)
+instrument_map.plot(axes=ax, title=False)
+instrument_map.draw_grid(axes=ax, color='k')
+for edge in edge_coords:
+    ax.plot_coord(edge, color='k', ls='-', marker='')
+ax.plot_coord(box_origin, color='r', marker='x')
+ax.plot_coord(hcc_orientation, color='b', marker='x')
+plt.show()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:138]:
```suggestion
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+#############################################################################################
+# Let's combine all of these pieces by plotting the lines of sight and the simulation box on
+# a single plot. We'll also overlay the pixel grid of our fake image.
+fig = plt.figure()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:143]:
```suggestion

fig = plt.figure()
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+          ymax=ypix_edges[-1],
+          color='k', ls='--', lw=.5, label='pixel grid')
+ax.hlines(y=ypix_edges,
+          xmin=xpix_edges[0],
+          xmax=xpix_edges[-1],
+          color='k', ls='--', lw=.5,)
+ax.legend(loc=2, frameon=False)
+ax.set_xlim(-10, 35)
+ax.set_ylim(-10, 35)
+plt.show()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:168]:
```suggestion
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+# falls in the box.
+
+def is_coord_in_box(box_edges, coord):
+    box_edges = SkyCoord(box_edges)
+    coord_hcc = coord.transform_to(box_edges.frame)
+    in_x = np.logical_and(coord_hcc.x<box_edges.x.max(), coord_hcc.x>box_edges.x.min())
+    in_y = np.logical_and(coord_hcc.y<box_edges.y.max(), coord_hcc.y>box_edges.y.min())
+    in_z = np.logical_and(coord_hcc.z<box_edges.z.max(), coord_hcc.z>box_edges.z.min())
+    return np.all([in_x, in_y, in_z], axis=0)
+
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:183]:
```suggestion


```
Maybe?

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+    inside_box = is_coord_in_box(edge_coords, los)
+    if inside_box.any():
+        ax.plot_coord(los[inside_box], color='C1', marker='', ls='-')
+for edge in edge_coords:
+    ax.plot_coord(edge, color='k', ls='-')
+ax.plot_coord(box_origin, marker='x', color='r')
+ax.plot_coord(hcc_orientation, color='b', marker='x', ls='')
+ax.set_xlim(-10, 35)
+ax.set_ylim(-10, 35)
+plt.show()
```
Comment (nabobalis) [inline examples/units_and_coordinates/los_simulation_box_intersection.py:213]:
```suggestion

plt.show()
```

File: examples/units_and_coordinates/los_simulation_box_intersection.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+hcc_orientation = SkyCoord(lon=0*u.deg, lat=20*u.deg,
+                           radius=sunpy.sun.constants.radius,
+                           frame='heliographic_stonyhurst')
+date = astropy.time.Time('2020-01-01')
+frame_hcc = Heliocentric(observer=hcc_orientation, obstime=date)
+
+#############################################################################################
+# Next, we need to define a coordinate frame for our synthetic spacecraft.
+# This could be something like SDO or *Solar Orbiter*, but in this case we'll just choose an
```
[the remark this answers is not in the document]
Comment (wtbarnes) [reply examples/units_and_coordinates/los_simulation_box_intersection.py:40]: Typically when you write out the full name of an observatory, it is in italics.

[the remark this answers is not in the document]
Comment (nabobalis) [reply examples/units_and_coordinates/los_simulation_box_intersection.py:40]: Oh woops, TIL

**Review by nabobalis — APPROVED** [review]

**Comment by wtbarnes** [comment]
This should go in the showcase section, not coordinates.

**Pull request merged**